Feature: Search Posts and Pages lists on WP.com/Jetpack sites

tonyr59h opened the pull request:
Fixes #62! I thought this was my oldest ticket fixed, but that honor belongs to #53. This PR will allow users to search their .com and Jetpack sites posts and pages. It's a long-requested feature so I think a lot of users will enjoy this.

Unfortunately self-hosted sites don't have an endpoint we can use to search remotely. I've left a note in the FluxC implementation to add support for self-hosted sites that have the WordPress REST API plugin available.

Analytics aren't included at the moment. I've opened #5729 to track this.

~A note about this PR: I noticed that the `PostsListAdapter` was handling a lot of work that should be delegated to the Fragment or Activity. I took the liberty of moving FluxC dispatch calls out, but there's still some loading being done that I think we should offload. That's for another PR though, I don't want to overcomplicate this feature.~

To test:
* Sign into a self-hosted site
* Verify that search is not available in the posts and pages lists
* Sign into a Jetpack and/or .com site
* Verify that search is available in the posts and pages lists
* Begin typing and verify that the progress indicator is displayed
* Wait for results (1.5s delay in searching) and verify progress indicator is hidden
* Type more before search results are shown to verify network calls are continuously delayed until user stops typing for 1.5s

cc @oguzkocer, I owe you some reviews by now 😉

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostsListActivity.java:
```diff
@@ -81,32 +86,33 @@ protected void onActivityResult(int requestCode, int resultCode, Intent data) {
+                    return true;
+                }
+            });
+            ((SearchView) mSearchAction.getActionView()).setOnQueryTextListener(this);
```
This is a little bit of a nitpick, but could we move the `(SearchView) mSearchAction.getActionView()` out of the `if` statement, so it doesn't get called twice. (The other one is inside the `if` expression)

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostsListActivity.java:
```diff
@@ -20,49 +23,51 @@
         ActionBar actionBar = getSupportActionBar();
         if (actionBar != null) {
-            actionBar.setTitle(getString(mIsPage ? R.string.pages : R.string.posts));
+            setTitle(getString(mIsPage ? R.string.pages : R.string.posts));
```
Another nitpick: It looks like we are calling this exact same `setTitle` twice. Do you think it'd be better to remove the `public void setTitle(CharSequence title) {` override and add a helper method for setting the title?

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostsListActivity.java:
```diff
@@ -122,5 +128,81 @@ public boolean onOptionsItemSelected(final MenuItem item) {
+        if (intent.hasExtra(EXTRA_VIEW_PAGES)) {
+            mIsPage = intent.getBooleanExtra(EXTRA_VIEW_PAGES, false);
+        }
+        if (intent.hasExtra(EXTRA_SEARCH_TERM)) {
```
I don't think we should be checking the `intent` for this extra, and instead should retrieve it from the `savedInstanceState` if present. As a consequence, if you rotate the device, the search term disappears.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostsListActivity.java:
```diff
@@ -122,5 +128,81 @@ public boolean onOptionsItemSelected(final MenuItem item) {
+    private PostsListFragment getPostListFragment() {
+        PostsListFragment frag = (PostsListFragment) getFragmentManager().findFragmentById(R.id.postList);
+        if (frag == null) {
+            frag = PostsListFragment.newInstance(mSite);
```
I am really not sure if we should be doing this. The newly created fragment is not getting attached to the view, so it doesn't seem like this is helping. Since the fragment is added in the xml and never removed, this check should never go through, right?

tonyr59h replied:
> this check should never go through, right?

Not sure why this was included but you're right, it should be safe without it (and it's running fine without it).

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostsListActivity.java:
```diff
@@ -122,5 +128,81 @@ public boolean onOptionsItemSelected(final MenuItem item) {
+    }
+
+    @Override
+    public boolean onQueryTextSubmit(String query) {
```
I am not sure if this is the reason yet, but if you write something in the search field and immediately submit, no search seems to be going through. Maybe we should be handling that here?

tonyr59h replied:
Thanks for testing that! There's no `mPostList.filterPosts(query);` call here, that's why the search doesn't go through.

tonyr59h replied:
Fixed in 3d0fb58e853895cf9d9daa245e5359294767c521

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostsListActivity.java:
```diff
@@ -20,49 +23,51 @@
     private SiteModel mSite;
 
+    private MenuItem mSearchAction;
+    private String mCurrentSearch;
```
I am not really sure why we have this `String mCurrentSearch` here, it doesn't look like anything is setting it.

Maybe you were planning to set it inside the `public boolean onQueryTextChange(String query)`?

tonyr59h replied:
This was going to be used to save/restore state, which you caught in another comment.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostsListFragment.java:
```diff
@@ -491,6 +452,33 @@ public void onPostButtonClicked(int buttonType, PostModel post) {
+        }
+    }
+
+    private void updateAdapter(List<PostModel> posts) {
```
We probably should rename this to something related to search since it's only used for that and the `updateAdapter` name is too generic that I was wondering why we weren't using it elsewhere.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostsListFragment.java:
```diff
@@ -608,7 +596,21 @@ public void onPostChanged(OnPostChanged event) {
                 setRefreshing(false);
                 hideLoadMoreProgress();
                 if (!event.isError()) {
-                    mCanLoadMorePosts = event.canLoadMore;
```
What happened to the mCanLoadMorePosts? How are we retrieving the other pages?

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostsListFragment.java:
```diff
@@ -420,44 +419,6 @@ public void onDetach() {
-     * called by the adapter to load more posts when the user scrolls towards the last post
-     */
-    @Override
-    public void onLoadMore() {
```
How are we triggering the fetch for the next set of posts/pages if we remove this?

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/posts/adapters/PostsListAdapter.java:
```diff
@@ -294,21 +275,27 @@ public void onClick(View v) {
-                if (mOnPostSelectedListener != null) {
-                    mOnPostSelectedListener.onPostSelected(post);
-                }
+                // TODO, move this logic to activity or fragment
```
It doesn't look like this is handled, so the post clicks don't do anything.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/posts/adapters/PostsListAdapter.java:
```diff
@@ -768,17 +743,11 @@ protected void onPostExecute(Boolean result) {
                         mediaToDownload.setMediaId(mediaId);
                         mediaToDownload.setLocalSiteId(mSite.getId());
-                        MediaPayload payload = new MediaPayload(mSite, mediaToDownload);
-                        mDispatcher.dispatch(MediaActionBuilder.newFetchMediaAction(payload));
```
Why are we not downloading the media anymore? I am a little confused.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostsListFragment.java:
```diff
@@ -641,6 +655,20 @@ public void onPostUploaded(OnPostUploaded event) {
 
+    @SuppressWarnings("unused")
+    @Subscribe(threadMode = ThreadMode.MAIN)
+    public void onPostsSearched(OnPostsSearched event) {
```
How are we handling the pagination for search here?

tonyr59h commented:
Thanks for the review! I'll patch it up, remove the refactor, and let you know when it's up for round 2 👍

tonyr59h commented:
Pagination added for searching and fixed for regular list browsing. Also fixed a couple issues; one with clicking on the post in search mode and another with cleaning up search results when changing the search term.

This is ready for another review @oguzkocer 😃

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostsListFragment.java:
```diff
@@ -633,6 +703,18 @@ public void onPostChanged(OnPostChanged event) {
+    private final Runnable mSearchRunnable = new Runnable() {
+        @Override
+        public void run() {
+            SearchPostsPayload payload = new SearchPostsPayload(mSite, mSearchTerm, mSearchOffset);
```
I know we have some checks beforehand, but do you think it makes sense to check if the `mSearchTerm` is empty here, especially to avoid some race conditions?

tonyr59h commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostsListActivity.java:
```diff
@@ -122,5 +133,66 @@ public boolean onOptionsItemSelected(final MenuItem item) {
+    public boolean onQueryTextSubmit(String query) {
+        mPostList.filterPosts(query);
+        setActionBarTitle("'" + query + "'");
+        return false;
```
D'oh! Good catch

tonyr59h commented:
> The drafts are not showing up when searched.

The default behavior of the [REST API](https://developer.wordpress.com/docs/api/1.1/get/sites/%24site/posts/) is to only search published posts. I'll add a ticket to FluxC to add search "any" status.

tonyr59h commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostsListFragment.java:
```diff
@@ -491,6 +515,37 @@ public void onPostButtonClicked(int buttonType, PostModel post) {
+    private void showSearchResults() {
+        if (mSearchResults.isEmpty()) {
+            mRecyclerView.setAdapter(null);
+            updateEmptyView(EmptyViewMessageType.NO_CONTENT);
```
Fixed in ae1f2e291748413b603562b7dce40d4bb18fd6da

tonyr59h commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostsListFragment.java:
```diff
@@ -609,6 +664,21 @@ public void onPostChanged(OnPostChanged event) {
                 hideLoadMoreProgress();
                 if (!event.isError()) {
                     mCanLoadMorePosts = event.canLoadMore;
+                    int count;
```
I think we can safely do without these changes in `onPostChanged`. They should have been removed when `onPostsLoaded` was brought back.

tonyr59h commented:
> Make sure you have more than 20 posts and keep scrolling the post list (make sure another page is loaded), go into search mode and type something, close the search mode and the 2nd page is now gone

Fixed in 380563f9a638871fd2f0a0aee47cc1843dd8a265.

I can't reproduce the first two bullet points, maybe they've been fixed in the changes since.

> The drafts are not showing up when searched.

wordpress-mobile/WordPress-FluxC-Android#452 adds support for searching posts of any status.

I'm fixing up the refresh on changed issue right now then I think that's all the kinks worked out.

tonyr59h commented:
Alright @oguzkocer, all issues have been addressed and this one is ready to go 👍

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostsListFragment.java:
```diff
@@ -87,6 +103,14 @@
+    private PostsListAdapter mSearchListAdapter;
+    private String mSearchTerm;
+    private Handler mHandler;
+    private boolean mIsSearching;
```
We need to save all these `mIsSearching`, `mCanSearchMore`, `mSearchOffset` in the bundle if we don't want to lose track of them on config changes (rotation etc).

tonyr59h commented:
Re-opening on a vastly simplified branch.
